src/persist/src/indexed/columnar.rs

```diff
@@ -351,104 +265,61 @@ impl<'a> ExactSizeIterator for ColumnarRecordsIter<'a> {}
 
 /// An abstraction to incrementally add ((Key, Value), Time, i64) records
 /// in a columnar representation, and eventually get back a [ColumnarRecords].
+#[derive(Debug)]
 pub struct ColumnarRecordsBuilder {
     len: usize,
-    key_data: Vec<u8>,
-    key_offsets: Vec<i32>,
-    val_data: Vec<u8>,
-    val_offsets: Vec<i32>,
+    key_data: BinaryBuilder,
+    val_data: BinaryBuilder,
     timestamps: Vec<i64>,
     diffs: Vec<i64>,
 }
 
-impl fmt::Debug for ColumnarRecordsBuilder {
-    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
-        fmt::Debug::fmt(&self.borrow(), fmt)
-    }
-}
-
 impl Default for ColumnarRecordsBuilder {
     fn default() -> Self {
-        let mut ret = ColumnarRecordsBuilder {
+        ColumnarRecordsBuilder {
             len: 0,
-            key_data: Vec::new(),
-            key_offsets: Vec::new(),
-            val_data: Vec::new(),
-            val_offsets: Vec::new(),
+            key_data: BinaryBuilder::new(),
+            val_data: BinaryBuilder::new(),
             timestamps: Vec::new(),
             diffs: Vec::new(),
-        };
-        // Push initial 0 offsets to maintain our invariants, even as we build.
-        ret.key_offsets.push(0);
-        ret.val_offsets.push(0);
-        debug_assert_eq!(ret.borrow().validate(), Ok(()));
-        ret
+        }
     }
 }
 
 impl ColumnarRecordsBuilder {
+    /// Reserve space for the given number of items with the given sizes in bytes.
+    /// If they end up being too small, the underlying buffers will be resized as usual.
+    pub fn with_capacity(items: usize, key_bytes: usize, val_bytes: usize) -> Self {
+        let key_data = BinaryBuilder::with_capacity(items, key_bytes);
+        let val_data = BinaryBuilder::with_capacity(items, val_bytes);
+        let timestamps = Vec::with_capacity(items);
+        let diffs = Vec::with_capacity(items);
+        Self {
+            len: 0,
+            key_data,
+            val_data,
+            timestamps,
+            diffs,
+        }
+    }
```

Comment:
As-is I think this will create a `ColumnarRecordsBuilder` that is not backed by LgAlloc, whereas it previously used to be?

Reply:
> As-is I think this will create a ColumnarRecordsBuilder that is not backed by LgAlloc

True!

> whereas it previously used to be?

`ColumnarRecordsBuilder` has never been backed by lgalloc as far as I know! Before this PR it was just a bunch of `Vec`s.

Reply:
Ooof, yeah you're totally right, I got it confused with the `ColumnarRecords` struct